### PR TITLE
Article Flag support for styles

### PR DESCRIPTION
**Pull request by killanaca**
ref #172 

- updated design
- fixed a warning on a missing title prop for the base ArticleFlag (inside the tests)
- use `Stylesheet` for internal component styling instead of plain object to avoid duplication on style attribute
- support external styling on the container and on the text element to avoid extra `<View>` to style the component on the `Article` component

```jsx
<View style={styles.ArticleFlagContainer}>
    <NewArticleFlag />
</View>
```
```jsx
const styles = StyleSheet.create({
  container: {
    backgroundColor: "#EAFF00"
  },
  text: {
    fontSize: 12
  }
});
<NewArticleFlag containerStyle={styles.container} textStyle={styles.text}/>
```

### CODE REVIEW

File: packages/article-flag/article-flag.js [hunk 1]
```diff
@@ -1,12 +1,12 @@
...
-import { View, Text } from "react-native";
+import { View, Text, StyleSheet } from "react-native";
 import PropTypes from "prop-types";
 
 import Diamond from "./diamond";
 
 // When changing styles please debug both web, android and ios because
 // some styles are not working correctly on all platforms (namely, android)
-const styles = {
+const styles = StyleSheet.create({
```
Comment (sericaia) [inline packages/article-flag/article-flag.js]: We were not using Stylesheet.create here for performance reasons

File: packages/article-flag/article-flag.js [hunk 2]
```diff
@@ -20,24 +20,24 @@ const styles = {
...
-};
+});
 
 // apply transformations to add uppercase and letter spacing.
 // letterSpacing CSS prop does not work on android:
 // https://github.com/facebook/react-native/pull/13199
 const beautifyTitle = title => title.toUpperCase().split("").join(" ");
 
-const ArticleFlag = ({ title, color }) => {
+const ArticleFlag = ({ title, color, containerStyle, textStyle }) => {
```
Comment (thatportugueseguy) [inline packages/article-flag/article-flag.js]: should style be a single prop that gets an object?

This way it keeps in line with html convention and it's normally used like that. I implemented it like this on byline: [link](https://github.com/newsuk/times-components/blob/master/packages/article-byline/article-byline-proptypes.js#L13)

File: packages/article-flag/article-flag.js [hunk 3]
```diff
@@ -46,20 +46,25 @@ const ArticleFlag = ({ title, color }) => {
 
 ArticleFlag.propTypes = {
   title: PropTypes.string.isRequired,
-  color: PropTypes.string
+  color: PropTypes.string,
+  containerStyle: PropTypes.number,
```
Comment (thatportugueseguy) [inline packages/article-flag/article-flag.js]: if we don't use the StyleSheet.create, you can use Text.proptype.style and View.proptypes.sytle

File: packages/article-flag/article-flag.js (hunk 1 above)
Comment (killanaca) [reply packages/article-flag/article-flag.js]: StyleSheet.create avoid to duplicate each style using ref instead

File: packages/article-flag/article-flag.js [hunk 4]
```diff
@@ -46,20 +46,23 @@ const ArticleFlag = ({ title, color }) => {
...
 
-const NewArticleFlag = () => <ArticleFlag title="new" color="#E34605" />;
-const UpdatedArticleFlag = () =>
-  <ArticleFlag title="updated" color="#3C81BE" />;
-const ExclusiveArticleFlag = () =>
-  <ArticleFlag title="exclusive" color="#C51D24" />;
-const SponsoredArticleFlag = () =>
-  <ArticleFlag title="sponsored" color="#4D4D4D" />;
+const NewArticleFlag = props =>
+  <ArticleFlag title="new" color="#E34605" {...props} />;
```
Comment (craigbilner) [inline packages/article-flag/article-flag.js]: any need to spread the props here?
also could just export each line...

File: packages/article-flag/article-flag.js [hunk 5]
```diff
@@ -17,23 +17,23 @@ const styles = {
...
-};
+});
 
 // apply transformations to add uppercase and letter spacing.
 // letterSpacing CSS prop does not work on android:
 // https://github.com/facebook/react-native/pull/13199
 const beautifyTitle = title => title.toUpperCase().split("").join(" ");
 
-const ArticleFlag = ({ title, color }) => {
+const ArticleFlag = ({ title, color, containerStyle }) => {
```
Comment (thatportugueseguy) [inline packages/article-flag/article-flag.js]: are we using containerStyle instead of style?

File: packages/article-flag/article-flag.js [hunk 6]
```diff
@@ -17,23 +17,23 @@ const styles = {
...
   if (!title) {
     return null;
   }
 
   return (
-    <View style={styles.view}>
+    <View style={[styles.view, containerStyle]}>
       <View style={styles.diamond}>
```
Comment (thatportugueseguy) [inline packages/article-flag/article-flag.js]: Do we really need the view around `Diamond`?

File: packages/article-flag/article-flag.js (hunk 4 above)
Comment (killanaca) [reply packages/article-flag/article-flag.js]: in this case I've just preferred to be more open on the props validation for a better bundle size for the web. 
However I've updated both

File: packages/article-flag/article-flag.js (hunk 5 above)
Comment (killanaca) [reply packages/article-flag/article-flag.js]: done

File: packages/article-flag/article-flag.js (hunk 6 above)
Comment (killanaca) [reply packages/article-flag/article-flag.js]: removed

**Comment by fampinheiro** [comment]
What is the reason for exporting multiple components instead of receiving a prop that renders a set of predefined styles?

```
<NewArticleFlag {...props } />
<UpdatedArticleFlag {...props } />
<ExclusiveArticleFlag {...props} />
```
instead of 

```
<ArticleFlag {...props } type="new" />
<ArticleFlag {...props } type="updated" />
<ArticleFlag {...props } type="exclusive" />
```

Not blocking the merge, just wondering the reason behind it. I find the usage of the second approach easier to use and grasp.

File: packages/article-flag/article-flag.js [hunk 7]
```diff
@@ -46,26 +42,53 @@ const ArticleFlag = ({ title, color }) => {
 
 ArticleFlag.propTypes = {
   title: PropTypes.string.isRequired,
-  color: PropTypes.string
+  color: PropTypes.string,
+  style: PropTypes.number
```
Comment (craigbilner) [inline packages/article-flag/article-flag.js]: is this right?

Comment (fampinheiro) [reply packages/article-flag/article-flag.js]: You probably want to use [ViewPropTypes.style](https://github.com/facebook/react-native/blob/master/Libraries/Components/View/ViewPropTypes.js#L391)

`import { ViewPropTypes } from "react-native"`

Comment (killanaca) [reply packages/article-flag/article-flag.js]: not yet, we have to stick with `View.propTypes.style` until we'll update React

**Comment by killanaca** [comment]
AFAIN if we want to use StyleSheet, we can't be strict eg:
```
NewArticleFlag.propTypes = {
        style: PropTypes.shape({
            marginRight: PropTypes.number
        })
};
```

otherwise we get  "Invalid prop `style` of type `number` supplied to `NewArticleFlag`, expected `object`." error message.

StyleSheet pass through a number as a ref not a real style object.
@craigbilner any suggestion?

**Pull request closed without merging**